src/index/MetaDataHandler.h

```diff
@@ -0,0 +1,302 @@
+    // _________________________________________________
+    std::pair<Id, const FullRelationMetaData&> operator*() const {
+      // make sure that we do not conflict with the empty key
+      AD_CHECK(_id != size_t(-1));
+      return std::make_pair(_id, std::cref(*_it));
+    }
+
+    // _____________________________________________________________
+    Iterator(Id id, const typename ExtVec::const_iterator& it, const ExtVec* const vec): _id(id), _it(it), _vec(vec){}
+
+
+    // ________________________________________________
+    Iterator& operator++() {
+      ++_id;
+      ++_it;
+      goToNexValidEntry();
+
+      return *this;
+    }
+
+    // ___________________________________________________
+    void goToNexValidEntry() {
+      while(_it != _vec->cend() && (*_it) == emptyMetaData) {
+	++_id;
+	++_it;
+      }
+    }
+
+    // ________________________________________________
+    Iterator operator++(int) {
+      Iterator old(*this);
+      ++_id;
+      ++_it;
+      goToNexValidEntry();
+      return old;
+    }
+
+    // TODO: I don't think we need both directions
+    /*
+    // ___________________________________________________
+    void goToPreviousValidEntry() {
+      while(_it != _vec->cbegin() && (*_it) == emptyMetaData) {
+	--_id;
+	--_it;
+      }
+    }
+
+    // ________________________________________________
+    Iterator& operator--() {
+      --_id;
+      --_it;
+      goToPreviousValidEntry();
+      return *this;
+    }
+
+    // ________________________________________________
+    Iterator operator--(int) {
+      Iterator old(*this);
+      --_id;
+      --_it;
+      goToPreviousValidEntry();
+      return old;
+    }
+    */
+
+    // _______________________________________________
+    bool operator==(const Iterator& other) {
+      return _it == other._it;
+    }
+
+    // _______________________________________________
+    bool operator!=(const Iterator& other) {
+      return _it != other._it;
+    }
+
+   private:
+
+    
+    Id _id;
+    typename ExtVec::const_iterator _it;
+    const ExtVec* const _vec;
+    // TODO: this must be handed in as a parameter
+    // And put this empty metaData to the FullRelationMetaData class.
+    FullRelationMetaData emptyMetaData = FullRelationMetaData(-1, -1, -1, 1.0, 1.0, false, false);
+  };
+
+}
+
+// _____________________________________________________________________
+template <class M>
+class MetaDataWrapperMmap {
+ public:
+  using Iterator = ExtVecWrapperImpl::Iterator<M>;
+
+  // _________________________________________________________
+  MetaDataWrapperMmap() = default;
+
+  // _________________________________________________________________
+  void setup(size_t numEls, const string& filename) {
+    _vec.setupNewFile(numEls, emptyMetaData, filename);
+  }
+
+  // __________________________________________________________________
+  // TODO: mark that this is the read version, e.g by using
+  // an enum-Flag
+  void setup(const string& filename) {
```

Comment:
If we can do the file specific stuff outside `MetaDataWrapperMmap` we may be able to have it generic enough to be able to use `std::array` and `std::vector` right?